build: update for preview 11

chadoh opened the pull request:
- Pin soroban-cli to latest commit from https://github.com/stellar/soroban-tools/pull/973. This PR will be mergeable once a new quickstart image is released and we can verify that its tests pass.
- Update the contract's SDK version
  - **Copy in `abundance` source from `token` contract** at https://github.com/stellar/soroban-examples/pull/277. Note that I **did not author most of the changes in the `contracts/abundance` folder**. I _only_ copied in the changes originally authored by @leighmcculloch, then updated for the single tweak that this `abundance` contract already had: 1. comments on the `mint` method; 2. slightly modified `mint` behavior.
- New CLI's typescript-bindings-generated libraries export a `Contract` class, rather than a flat bag of functions. Instantiating these contracts is now taken care of in `shared/contracts.ts`, with code throughout the rest of the codebase referencing these instances.
- Instantiating these classes required access to the `network` and `rpcUrl` used with the `initialize.sh` script. NextJS doesn't allow reading the files we already write (like `.soroban-example-dapp/network`), so for now I've also `echo`d relevant values to a `shared/config.json` file, which is hidden. This could probably be cleaned up in the future by putting all relevant settings in an `.env` file, and cleaning up the package.json scripts to use `dotenv cross-env` or similar.

chadoh commented on components/molecules/deposits/index.tsx:
```diff
@@ -2,21 +2,20 @@ import React from 'react'
   decimals: number
   name?: string
   symbol?: string
-  idCrowdfund: string
```
This was already obsolete; we just missed removing it previously

sreuland commented:
@chadoh , can probably use the latest commit off quickstart:master of `b58a1127ada30a477f49f330e21038426fb272e4` , also, the network config here could now allow running against `testnet` as well since it's on soroban/p20.

Julian-dev28 commented:
Just tried this out. Everything is working an no issue to report!

leighmcculloch reviewed (approved):
👏🏻

Julian-dev28 commented:
> Just tried this out. Everything is working an no issue to report!

I want to point out that I initially tried this out on _futurenet_  which worked with no issues to report. However, after trying it again on standalone, I ran into the error: 'xdr processing error: xdr value invalid.'

sreuland commented on .cargo/config.toml:
```diff
@@ -1,7 +1,7 @@
 
 [alias] # command aliases
-install_soroban = "install --git https://github.com/stellar/soroban-tools --rev cb3c44f9d8080917a7cb019d6be25019f6cf78c3 --root ./target soroban-cli --debug"
+install_soroban = "install --git https://github.com/AhaLabs/soroban-tools --rev dc2a543993a293155516df52e79cc120cbd3dfe0 --root ./target soroban-cli --debug"
```
can this use `https://github.com/stellar/soroban-tools --rev 20.0.0-rc2` or is there a branch on your fork that needs to be merged back to stellar/soroban-tools?

chadoh replied:
yes, there's a branch (https://github.com/stellar/soroban-tools/pull/973), see PR description.

This branch is waiting on a quickstart that works with preview 11 + standalone.

The lack of such a quickstart is probably also the cause of your troubles running this on standalone.

sreuland replied:
i see, you are referring to the `stellar/quickstart:soroban-dev` docker image? yes, it's been recently published for preview11 as [stellar/quickstart:testing@sha256:1c98f895f8b69cc843eeaa5230d67044dbeb390a5529d51dd7762d8ff685c3f8](https://soroban.stellar.org/docs/reference/releases#preview-11-september-11-2023-testnet-and-futurenet-edition)

sreuland commented:
> @chadoh , can probably use the latest commit off quickstart:master of `b58a1127ada30a477f49f330e21038426fb272e4` , also, the network config here could now allow running against `testnet` as well since it's on soroban/p20.

I [replied here](https://github.com/stellar/soroban-example-dapp/pull/136#discussion_r1333343676) for the quickstart docker version which you were probably more interested in.

on the 'testnet' enablement, that would be adding it as another network choice in initialize.sh and quickstart.sh.

also, in regards to using the preview11 quickstart version, it has deprecated the usage of `--standalone` in favor of `--local` instead now, it will print a warning for the former usage. so, it may be good to adjust terminology here to be consistent, i.e. replace 'standalone' with 'local' or should we carve that out to separate ticket?

chadoh commented:
I do think that changing all the language from `quickstart` to `local` or `localnet` is a separate issue. Partly because I'm not clear on what the new name is. Just `local`? Or `localnet`? Or does it depend on context? Seems like a bike-shed-type change that doesn't need to slow this one down.

I would also prefer the addition of `testnet` to the options here be a separate PR, too. Is that ok? This one was just meant to get things working with the latest Futurenet, as quickly as possible.

I updated the quickstart image in https://github.com/stellar/soroban-tools/pull/973 and here, and I switched the `--standalone` flag to `--local` in the `quickstart.sh` script. However, you can see that the change to the CLI has a [failing build](https://github.com/stellar/soroban-tools/actions/runs/6291027336/job/17079035047):

    Error: Failed to initialize container stellar/quickstart:testing@sha256:1c98f895f8b69cc843eeaa5230d67044dbeb390a5529d51dd7762d8ff685c3f8

sreuland commented:
> I would also prefer the addition of testnet to the options here be a separate PR, too. Is that ok? This one was just meant to get things working with the latest Futurenet, as quickly as possible.

@chadoh , thanks sounds good.

> I updated the quickstart image in https://github.com/stellar/soroban-tools/pull/973 and here, and I switched the --standalone flag to --local in the quickstart.sh script. However, you can see that the change to the CLI has a [failing build](https://github.com/stellar/soroban-tools/actions/runs/6291027336/job/17079035047):

ok, will try out locally, we just need to verify initialize&run of dapp for local and futurenet network types correct?

sreuland commented:
@chadoh , that format error from gh workflow is a platform discrepancy of host and image, that sha is pointing at wrong image version actually, you can see earlier in the logs of it downloading the image it reports the issue `WARNING: The requested image's platform (linux/arm64) does not match `. 

So, that's one issue, which needs to get fixed in soroban docs also, as it should be the multi-platform sha instead and needs to be updated to newer build anyways, because, I'm trying the pr locally on an arm64 host for the standalone network, where that image gets past format errors, but errors for real on contract deploys during init due to resource limits on the local network from`quickstart:testing` image, the latest build of `stellar/quickstart:soroban-dev` actually has more recent changes to disable resource limits on that image per https://github.com/stellar/quickstart/pull/492, so, should grab that, and change `QUICKSTART_SOROBAN_DOCKER_SHA=stellar/quickstart:soroban-dev` and add `--pull always` to the docker run command in quickstart.sh to make sure getting the latest, this will also allow docker to pull the correct platform image automatically.

I've manually installed latest cli - `cargo install soroban-cli@20.0.0-rc4 --root ./target`

With that in place , i'm able to run `./quickstart.sh standalone` and then `NETWORK=standalone npm run reset` , init logs that it used my cli -`Using soroban binary from ./target/bin`  and contracts are built/deployed.

but when I run the server - `npm run dev` i get a js error in browser:

<img width="1149" alt="Screen Shot 2023-09-26 at 2 17 55 PM" src="https://github.com/stellar/soroban-example-dapp/assets/10703845/c32c3812-3c88-4ad1-bda6-feadf929147f">

chadoh commented:
@sreuland this one should be working now

sreuland commented on .cargo/config.toml:
```diff
@@ -1,7 +1,7 @@
 
 [alias] # command aliases
-install_soroban = "install --git https://github.com/stellar/soroban-tools --rev cb3c44f9d8080917a7cb019d6be25019f6cf78c3 --root ./target soroban-cli --debug"
+install_soroban = "install --git https://github.com/AhaLabs/soroban-tools --rev c7fb7e08ba8efa9828d9df863d991558f269e35b --root ./target soroban-cli --debug"
```
can this be changed now to use a released version of cli like:

```
install --root ./target soroban-cli@20.0.0-rc4 --debug
```

sreuland commented:
> @sreuland this one should be working now

@chadoh , I tried latest from pr, it's initializing correctly but the dapp in browser is getting different js error:

I did in one terminal:
```
./quckstart.sh standalone
```

and in second terminal:
```
rm -rf ./node-modules
npm install
NETWORK=standalone npm run reset
npm run dev
```

in browser i get:
<img width="926" alt="Screen Shot 2023-09-28 at 12 51 24 PM" src="https://github.com/stellar/soroban-example-dapp/assets/10703845/10eed5af-cadf-4d14-ae76-3ccb58bce651">

chadoh commented on quickstart.sh:
```diff
@@ -19,7 +19,7 @@ esac
 # this is set to the quickstart `soroban-dev` image annointed as the release 
 # for a given Soroban Release, it is captured on Soroban Releases - https://soroban.stellar.org/docs/reference/releases 
-QUICKSTART_SOROBAN_DOCKER_SHA=stellar/quickstart:soroban-dev@sha256:a6b03cf6b0433c99f2f799b719f0faadbb79684b1b763e7674ba749fb0f648ee
+QUICKSTART_SOROBAN_DOCKER_SHA=stellar/quickstart:soroban-dev
```
`quickstart:testing` or `quickstart:soroban-dev`?

sreuland replied:
it's the `quickstart:testing`, docker will resolve the same sha from both `docker.io/stellar/quickstart:testing@sha256:40636cdb1b9168b47e5dc120949fe3610ff914e8dd43409edb6fa66496bdd9c3` or `docker.io/stellar/quickstart@sha256:40636cdb1b9168b47e5dc120949fe3610ff914e8dd43409edb6fa66496bdd9c3`

sreuland replied:
@chadoh , one minor cleanup, can you set `QUICKSTART_SOROBAN_DOCKER_SHA=docker.io/stellar/quickstart:testing@sha256:40636cdb1b9168b47e5dc120949fe3610ff914e8dd43409edb6fa66496bdd9c3` , then this pr is done, and ready to merge both @Julian-dev28 and I have tested locally as well.

sreuland commented:
> @chadoh , I tried latest from pr, it's initializing correctly but the dapp in browser is getting different js error:

I think this error was due to the local 'target' directory on my machine not being wiped as part of dapp init steps so, some older contract was in there, and bindings got generated from that, or at least after purging the target directory and init'ing dapp from npm reset again, this forced new cli, ontract wasm, contract bindings from the wasm to be re-built regardless, and eliminated the error on ts gen code saw earlier, and the dapp loads now. 

During investigation, I tried following the steps in README for how to init, and it gets pretty complex with different instructions for different networks, so, I took a small step back and tried to consolidate dapp init to always just be `npm run reset`, to avoid any old state between dapp restarts, etc and a single way to init regardless of network type. 

I put up PR with these suggestions based on your `build/preview11` branch, in case you agree the suggestions and want to pull in. [AhaLabs/soroban-example-dapp/pull/1](https://github.com/AhaLabs/soroban-example-dapp/pull/1)

sreuland reviewed (approved):
@chadoh , looks good, I just tested out standalone and futurenet back to back, and both cases worked, the only thing I found was the 'clean' behavior between resets and put up a pr with a suggestion - [AhaLabs/soroban-example-dapp/1](https://github.com/AhaLabs/soroban-example-dapp/pull/1)

Julian-dev28 commented:
> @chadoh , looks good, I just tested out standalone and futurenet back to back, and both cases worked, the only thing I found was the 'clean' behavior between resets and put up a pr with a suggestion - [AhaLabs/soroban-example-dapp/1](https://github.com/AhaLabs/soroban-example-dapp/pull/1)

Just ran both cases back to back as well and can confirm this is working
